kernel/src/main/java/org/kframework/parser/concrete2kore/disambiguation/KAppToTermConsVisitor.java

```diff
@@ -0,0 +1,99 @@
+import scala.collection.immutable.Set$;
+import scala.util.Either;
+import scala.util.Left;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.kframework.Collections.*;
+import static org.kframework.kore.KORE.*;
+
+
+/**
+ * Transform the KApps found in a term into the corresponding TermCons so type checking and
+ * variable type inference takes place correctly. Must be applied between type inference and
+ * priority filter.
+ */
+public class KAppToTermConsVisitor extends SetsTransformerWithErrors<KEMException> {
+
+    private final Module mod;
+    public KAppToTermConsVisitor(Module mod) {
+        super();
+        this.mod = mod;
+    }
+
+    @Override
+    public Either<java.util.Set<KEMException>, Term> apply(TermCons tc) {
+        assert tc.production() != null : this.getClass() + ":" + " production not found." + tc;
+        if (tc.production().klabel().isDefined() && tc.production().klabel().get().name().equals("#KApply")) {
+            if (!(tc.items().get(1) instanceof Constant) || !((Constant) tc.items().get(1)).production().sort().equals(Sorts.KLabel()))
+                // TODO: maybe return a hidden warning?
+                return super.apply(tc); // don't do anything if the label is not a token KLabel (in case of variable or casted variable)
+            Constant kl = (Constant) tc.items().get(1);
+            PStack<Term> items = flattenKList(tc.items().get(0));
+            String klvalue = kl.value();
+            try { klvalue = StringUtil.unescapeKoreKLabel(kl.value()); } catch (IllegalArgumentException e) { /* ignore */ } // if possible, unescape
+            Set<Production> prods = mutable(mod.productionsFor().get(KLabel(klvalue))
+                    .getOrElse(Set$.MODULE$::emptyInstance)
+                    .filter(x -> ((Production) x).arity() == items.size()).toSet());
+            if (prods.size() == 0) {
+                String msg = "Could not find any production with arity " + items.size() + " for label " + kl.value();
+                return Left.apply(Sets.newHashSet(KEMException.innerParserError(msg, kl)));
+            } else if (prods.size() == 1)
+                return super.apply(TermCons.apply(items, prods.iterator().next(), tc.location(), tc.source()));
+            else {
+                // instantiate all labels found and let the type checker filter them out
+                Set<Term> tcs = new HashSet<>();
+                for (Production prd : prods)
+                    tcs.add(TermCons.apply(items, prd, tc.location(), tc.source()));
+                return super.apply(Ambiguity.apply(tcs, tc.location(), tc.source()));
+            }
+        }
+        return super.apply(tc);
+    }
+
+    /**  Recurse under #KList and flatten all the terms */
+    private static PStack<Term> flattenKList(Term t) {
+        if (t instanceof Ambiguity) {
+            Ambiguity amb = (Ambiguity) t;
```

Comment:
The logic in this block doesn't look sound to me. You are basically dropping some of the parses in the forest

Reply:
I am applying an early `prefer` filter, only for `#KList`. If I get a KApp, I gather all the KList items and put them in a single list. If I get different ambiguities, like user lists, I discard them since I'm only interested in KList in this place.
I want to apply this filter in exactly this place because it simplifies this part. It is possible to have user lists mixed in, then the number of KList items is going to be smaller. That's why I'm only taking the list with the highest amount of elements. Since KList has the left attribute, I'm only going to find one element with the highest amount of items.

It's possible that I have a production with arity 2, but a KList with 3 elements. Then that's going to be an error. You can still write that if you add a set of braces (see test:495)

If I make it the first filter, I have to handle more cases when flattening KList. It could also hinder priorities. I can write `mul(plus(1,2),3)` but when I transform it into TermCons, it's going to be rejected by the priority filter.